### ISSUE TITLE
Suggestion

### ISSUE THREAD

**Issue by YHL04** [lucidrains/block-recurrent-transformer-pytorch#5]
Can you try to input the recurrent state to the first transformer layer? Like a recurrent layer but without next state? Maybe that will improve the performance since high level representations don't have to be recomputed and remembered info can be passed through the whole model again without any additional compute cost

**Comment by lucidrains** [comment]
@YHL04 yea, that may be a good idea, let me think about how to adjust the code for this

are you running any experiments on this?

**Comment by lucidrains** [comment]
you basically mean reading from the state at a lower layer, and writing to the state at a higher representation layer?

**Comment by lucidrains** [comment]
would you be willing to run and share some experiments, if I were to get this feature built?

**Comment by lucidrains** [comment]
@YHL04 actually its ok, i'm running the experiments this morning. i need to know

write code, run experiments, read papers, choose 2 ...

**Comment by YHL04** [comment]
@lucidrains I might run a few experiments although my implementation is not that accurate according to paper, i also don't have that much compute. I meant like detaching the previous state and adding it to the first layer without changing anything else, although if we were to input the state at an earlier layer and output the state at a later layer it might or might not work, because the feedforward in the recurrent layer and lstm gate did not result in performance gain according to paper

**Comment by lucidrains** [comment]
oh, how would you add it to the first layer? like the first layer would cross attend to the previous states, but where would you compute the next states, in your modification?

**Comment by lucidrains** [comment]
yeah, the idea of passing higher level reps back to earlier layers had been explored in feedback transformer, which is why I thought it may be good to try in this context (in this paper, it is traversed block by block with the learned EMA gate)

**Comment by YHL04** [comment]
im not sure, probably using the recurrent mechanism but without next state, basically the recurrent layer is unchanged its just that the state is detached and added at the top... although changing the input and output recurrent state to different layers is possible but a lot more complicated

**Comment by lucidrains** [comment]
@YHL04 ahh got it

yea, i'll proceed with the ability to decouple which layers recurrent state is read and written to

**Comment by YHL04** [comment]
@lucidrains how'd it go

**Comment by YHL04** [comment]
i did some experiments but the loss curve is kinda identical

**Comment by lucidrains** [comment]
@YHL04 hey, i just completed this feature this morning, and upon running the experiments... i don't see much of anything either :laughing:

oh well, i'll leave the feature around (`read_recurrent_layers` hyperparameter) for others to explore

**Comment by JDvorak** [comment]
I know this is closed, but it seems like a variation of this line of thought could be used to bridge the gap with [Continual Transformers](https://github.com/LukasHedegaard/continual-inference) and enable Block-Recurrent Transformers to operate as also a streaming transformer.

**Comment by YHL04** [comment]
Have you ever thought about caching the gradients for the recurrent state to enable it to train for much longer sequences while fitting into memory? Like BPTT 80 instead of 10

**Comment by lucidrains** [comment]
@YHL04 i'm actually a lot more bullish on this [recurrent transformer approach](https://github.com/lucidrains/recurrent-memory-transformer-pytorch). i have also brought over a memory checkpointing technique from the memformer paper, enabling very long sequence training without memory issues

**Comment by lucidrains** [comment]
@YHL04 i think the block recurrent transformer is a bit overengineered
